ui/app/AppLayouts/Profile/popups/ChangePasswordModal.qml

```diff
@@ -12,113 +12,163 @@ import StatusQ.Popups 0.1
+            onStateChanged: if (state == "init") resetInternal()
+
+            states: [
+                State {
+                    name: "init"
+                },
+                State {
+                    name: "typing"
+                    PropertyChanges { target: currentPasswordInput; validationError: d.lengthValidator(text) }
+                },
+                State {
+                    name: "incorrect"
+                    PropertyChanges { target: currentPasswordInput; validationError: qsTr("Incorrect password") }
+                }
+            ]
         }
 
-        //TODO replace with StatusInput as soon as it supports password
+        // TODO replace with StatusInput as soon as it supports password
         Input {
             id: passwordInput
+
+            readonly property bool ready: state == "typing" && validationError == ""
+
             anchors.left: undefined
             anchors.right: undefined
             Layout.fillWidth: true
+            label: qsTr("New password")
+
+            textField.echoMode: TextInput.Password
+            keepHeight: true
             placeholderText: ""
-            label: qsTrId("new-password...")
+            state: "init"
+
+            onTextChanged: if (text != "" && state != "typing") state = "typing"
+            onStateChanged: if (state == "init") resetInternal()
+
+            states: [
+                State {
+                    name: "init"
+                },
+                State {
+                    name: "typing"
+                    PropertyChanges { target: passwordInput; validationError: d.lengthValidator(text) }
+                }
+            ]
+        }
+
+        // TODO replace with StatusInput as soon as it supports password
+        Input {
+            id: confirmPasswordInput
+
+            readonly property bool ready: state == "typing" && validationError == ""
+
+            anchors.left: undefined
+            anchors.right: undefined
+            Layout.fillWidth: true
+            label: qsTr("Confirm new password")
+
             textField.echoMode: TextInput.Password
-            onTextChanged: {
-                popup.passwordInputValid = ((passwordInput.text !== "") && (passwordInput.text.length >= 6));
-                //setting validationError so that input becomes red
-                passwordInput.validationError = (!popup.passwordInputValid) ? " " : "";
-                popup.indicationText = (!popup.passwordInputValid ? "<font color=\"#FF2D55\">" : "")
-                    + "At least 6 characters." + (!popup.passwordInputValid ? "</font>" : "")
-                    + "Your password protects your keys. You need it to unlock Status and transact."
-            }
+            keepHeight: true
+            placeholderText: ""
+            state: "init"
+
+            onTextChanged: if (text != "" && state != "typing") state = "typing"
+            onStateChanged: if (state == "init") resetInternal()
+
+            states: [
```

Comment:
can the states be combined to a common for the whole file and all the 3 properties being affected can be changed from there?

Reply:
Hey I just saw "init", "typing" where common thought they could be merged but off course doesn't need to be right now. approving it for now :)

Reply:
Initially, this whole logic was extracted to a separate Component: https://github.com/status-im/status-desktop/blob/83854ddd9be940a2344e6f1ae125f1baeb955d8d/ui/imports/shared/controls/PasswordInput.qml
But it was reverted due to: https://github.com/status-im/status-desktop/pull/4400#discussion_r784170200 which unfortunately results in code duplication

The states must not be combined to a common one, each `Input` should have its own state as they are separated entities.